cli,server: produce JDBC-compatible connection URLs

knz opened the pull request:
Fixes #63962. 
Fixes #35998.
cc @cockroachdb/server 

This PR contains a batch of changes related to the production and handling of connection strings (URLs) inside CockroachDB.

Salient changes:

- ensure a database name is always included in the URL (better compatibility with JDBC)
- new sub-command `cockroach convert-url` to convert a connection URL between libpq and JDBC format (example below)
- the libpq-compatible and JDBC URLs are now printed side-by-side by `cockroach start` and `cockroach demo`.

Example output for `cockroach convert-url`:
![image](https://user-images.githubusercontent.com/642886/119242378-79845380-bb5d-11eb-8599-fd4aa643ccd4.png)

Example output for `cockroach start`:
```
sql:                 postgresql://root@kenax:26257/defaultdb?sslmode=disable
sql (JDBC):          jdbc:postgresql://kenax:26257/defaultdb?sslmode=disable&user=root
```

Example output for `cockroach demo`:

```
#   (sql)      postgresql://demo:demo64277@127.0.0.1:26257/movr?sslmode=require
#   (sql/jdbc) jdbc:postgresql://127.0.0.1:26257/movr?password=demo64277&sslmode=require&user=demo
```

knz committed:
cli/demo: keep the demo database name across reconnects

Release note (bug fix): The URLs printed out by the client-side
command `\demo ls` in `cockroach demo` now properly include
the workload database name, if any was created.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65460)
<!-- Reviewable:end -->

bdarnell commented:
Neat - I was just testing this and didn't realize that jdbc supported username/password in query parameters (I assumed the lack of support for the `user:password@` was them taking a stance against putting passwords in urls at all). However, we need to test this to make sure it's supported by other drivers too. From a quick glance at the code I don't think that the javascript pg driver will work with this url. (I have a testbed for this and can confirm later). 

> This increases compatibility with more drivers and handles special
characters properly.

The handling of special characters varies by driver - python/psycopg2 is a stickler for correct url encoding even in query parameters where there is no ambiguity.

knz commented:
> The handling of special characters varies by driver - python/psycopg2 is a stickler for correct url encoding even in query parameters where there is no ambiguity.

My point is more that Go's `url.Encode` function doesn't properly urlencode characters in the password field before `@`, but does properly urlencode the query arguments.

rafiss commented on pkg/cli/demo_cluster.go:
```diff
@@ -883,29 +893,26 @@ func generateCerts(certsDir string) (err error) {
 	}
 	sqlURL := url.URL{
-		Scheme: "postgres",
+		Scheme: "postgresql",
```
would it be worth introducing a `const PostgreSQLScheme = "postgresql"` somewhere so that all these url builders can refer to the same thing?

(btw, i hope there aren't any drivers out there that only accept `postgres` and not `postgresql`. i haven't been able to find one yet)

knz replied:
Done.

rafiss reviewed (commented):
oh,, just saw Ben's comment. confirmed that with the `node-postgres` driver (JS) the connection string `postgresql://localhost:26257/defaultdb?sslmode=disable&user=root`

results in 
```
error: password authentication failed for user rafiss
    at Connection.parseE (/Users/rafiss/scratch/node_modules/pg/lib/connection.js:567:11)
    at Connection.parseMessage (/Users/rafiss/scratch/node_modules/pg/lib/connection.js:391:17)
```
(username is not picked up)

knz commented:
@rafiss can you try `postgresql://root@localhost:26257/defaultdb?sslmode=disable&user=root` ?

i.e. pass the username both before @ and in the query args

knz commented:
looking at the node-postgres docs I think it won't work. maybe we need the two urls side-by-side.

rafiss commented:
confirmed -- `postgresql://root@localhost:26257/defaultdb?sslmode=disable&user=root` works with `node-postgres`

> maybe we need the two urls side-by-side.

I think this might be the best approach for resolving this issue -- show the existing URL and a "JDBC friendly URL." though for both of these URLs, using a scheme of `postgresql` instead of `postgres` seems better

---

this may be an aside, but one more bit of trivia to be aware of: some ORMs use the URL scheme to decide which _ORM_ adapter/dialect to use. for example, sqlalchemy and typeorm need you to use a URL like `cockroachdb://root@localhost:26257/defaultdb?sslmode=disable` in order for them to work properly with CRDB. (ORMs usually also usually let you configure each property one by one and not use a URL at all)

emily-horing commented:
@knz just making sure I understand correctly, this is adding another option for formatting the url to work better with certain tools, but using `username:password` would still work, correct? So for example if I already had a connection up between my app and CRDB I would not have to make any changes?

knz commented:
Emily, yes.

knz commented:
Emily, what this work is revealing though is that some client tools need one URL and some other tools need another. 

So for your work I think it will be useful to present two different URLs in the "connection string" tab in CC. The first one would be the current one, and the one under it would be "for Java/JDBC applications".

emily-horing commented:
Agreed, I think medium-term we need to iron out how everything we present in the UI can be just copy/pastable for the user with minimal input. For now we will rely a bit more on docs though and direct users to find the exact code snippets there as we figure out designs.

knz committed:
pgurl: new package to manage postgres connection strings

The new `pgurl` package contains helper functions to construct,
validate and convert postgres connection strings.

Release note: None

knz committed:
cli: new subcommand `cockroach convert-url`

Release note (cli change): The new subcommand `cockroach convert-url`
converts a connection URL, such as those printed out by
`cockroach start` or included in the online documentation,
to the syntax recognized by various client drivers.

For example:

```
$ ./cockroach convert-url --url "postgres://foo/bar"
Connection URL for libpq (C/C++), psycopg (Python), lib/pq & pgx (Go),
node-postgres (JS) and most pq-compatible drivers:

   postgresql://root@foo:26257/bar

Connection DSN (Data Source Name) for Postgres drivers that accept
DSNs - most drivers and also ODBC:

   database=bar user=root host=foo port=26257

Connection URL for JDBC (Java and JVM-based languages):

   jdbc:postgresql://foo:26257/bar?user=root
```

knz committed:
cli,sql: fix some import path ambiguities

Release note: None

knz committed:
cli,rpc,sql: use the new `pgurl` package where relevant

This simplifies the code in multiple places and makes it generally
easier to learn & maintain.

Additionally, certain drivers were confused by the lack of database
name between the hostname and the query parameter, e.g. in
'postgres://foo?arg=val`

This fixes that issue by ensuring there is always a database path
included.

Release note (bug fix): When a non-SQL CLI command (e.g. `cockroach
init`) was invoked with the `--url` flag and the URL did not contain
a `sslmode` parameter, the command was incorrecting defaulting
to operate as if `--insecure` was specified. This has been corrected.

Release note (cli change): the URLs spelled out by `cockroach start`,
`start-single-node`, `demo` in various outputs now always contain a
target database for the connection, for example `defaultdb` for
regular servers. Certain drivers previously automatically filled in
the name "postgres" if the database name field was empty.

knz committed:
cli/start,demo: produce JDBC-compatible connection URLs

Some client drivers do not recognize user/password information passed
before the "@" sign in the connection URL, most notably the JDBC
driver.

Example when a new server starts:

Before:
```
sql:        postgresql://demo:demo123@kenax:26257/defaultdb?sslmode=disable
```

After:
```
sql:        postgresql://demo:demo123@kenax:26257/defaultdb?sslmode=disable
sql (JDBC): postgresql://kenax:26257/defaultdb?sslmode=disable&user=demo&password=demo123
```

Release note (cli change): The connection URLs spelled out by
`cockroach start`, `start-single-node`, `demo` in various outputs now
include a variant suitable for use with JDBC client apps.

knz commented:
RFAL: See the updated PR description at the top for a summary.

I have reworked the PR entirely. There is now a new package `pgurl` that handles the various URL formats, consistently across all the parts of CockroachDB that use URLs. Both JDBC and libpq-compatible URLs are now displayed side-by-side.

knz committed:
cli/demo: simplify the socket URL code to use `pgurl`

Release note: None

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65460)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @emily-horing, and @knz)

___
*[pkg/server/pgurl/testdata/url, line 14 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65460#-Mab2HDf7ufUQDHAxiar:-Mab2HDgDje_-8gKf48_:bmtaqq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f2fb3a5af285e09aeab2d55d6b7d8e44a9bef2/pkg/server/pgurl/testdata/url#L14)):*
> ```
> # Empty host.
> url
> postgres:///somedb
> ```

do other tools really treat empty fields like this as valid? that's a bit surprising to me. it seems like it would be fine if we were more restrictive, but i suppose it's nice to support as much as we can as long as it's standard

(i had a similar thought for a case below that has a username but empty password)


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65460)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @emily-horing, and @rafiss)

___
*[pkg/server/pgurl/testdata/url, line 14 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65460#-Mab2HDf7ufUQDHAxiar:-MacmFLU5N6u98vo3i0D:b-gad8uh) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f2fb3a5af285e09aeab2d55d6b7d8e44a9bef2/pkg/server/pgurl/testdata/url#L14)):*
> it would be fine if we were more restrictive

that is possibly the case, but I'm coming to this from the other end - the omitted fields syntax was allowed by previous versions of cockroachdb, and I'm generally shy of removing functionality.

> a username but empty password

that, on the other hand, seems like a legitimate use case. Why do you find this surprising?


<!-- Sent from Reviewable.io -->

knz committed:
cli/flags.go: add an explanatory comment

Release note: None

knz committed:
cli/client_url: add validation details to error strings

This patch works around a limitation in the upstream
`pflag` package.

Release note: None

knz commented:
I have rebased this and fixed all problems that had been detected by tests (hopefully).

Is this PR still relevant? If so would appreciate another look.

rafiss reviewed (approved):
this looks good to me!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65460#-:-MbOK0Tm9VP6WuRCHu_1:b-2bm22q)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @emily-horing, and @knz)

___
*[pkg/server/pgurl/testdata/url, line 14 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65460#-Mab2HDf7ufUQDHAxiar:-MbOJYEM3gSS3ds5r5CZ:be4i4od) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f2fb3a5af285e09aeab2d55d6b7d8e44a9bef2/pkg/server/pgurl/testdata/url#L14)):*
>  the omitted fields syntax was allowed by previous versions of cockroachdb, and I'm generally shy of removing functionality.

makes sense

> username but empty password

i was just surprised that a username followed by a colon followed by nothing worked, but it does make sense as a use case


<!-- Sent from Reviewable.io -->

knz commented:
TFYR!

bors r=rafiss

knz commented:
build failed due to fastly outage

knz commented:
bors r=rafiss

knz commented:
@rafiss do we need this backported to 21.1?

rafiss commented:
The new URL info would be helpful. A backport would be nice, though it doesn't seem like these changes satisfy the "Patches should change as little code as possible" condition in our backport policy. So if you feel like the code changes are risky, then I am fine without the backport.

knz commented:
yes, if we want to backport this we'll need a higher level of scrutiny.

we can de-risk it by creating a smaller PR which only uses the new `pgurl` package to print out the URL but keeps the previous code alongside it (i.e. accept code duplication in 21.1).

Both options would require more work hours than a regular backport, so I'd rather only do it if we have clear demand.

rafiss commented:
Let's not backport then. Thanks for your work here!
